### PR TITLE
The Windows Update

### PR DESCRIPTION
**Pull request by gpotter2**
This PR includes a major refactor regarding several Windows specific protocols, particularily SMB2/3 and DCE/RPC. It also touches many parts of Scapy's core in order to accommodate those changes.

This includes:
  - DCE/RPC:
      - DCERPC_Client with support for NCACN_IP_TCP and NCACN_NP
      - DCERPC_Server with support for NCACN_IP_TCP and NCACN_NP
      - Finish NDR engine !
      - Server/Client Endpoint mapper support - Premises of a few special clients (Netlogon, DCOM, ...)
      - A few RPC interfaces (to debug/test the DCE/RPC engine. TODO: MORE !) - Documentation !
  - SMB2/3:
      - Protocol refactor, many more SMB2/3 structures supported
      - Server (class + 'simple' util smbserver()) (2.0.2 to 3.1.1) 
      - Client (class + interactive CLI smbclient()) (2.0.2 only) 
      - SMB socket, RPC over SMB socket, etc. 
      - Documentation !
  - Kerberos: 
      - KerberosSSP to use in SMB/RPC clients/servers 
      - Crypto: use cryptography, latest RFC8009, typing, etc. 
      - Util functions krb_as_req, krb_tgt_req, kpasswd (both modes), etc. 
      - [MS-KILE] variants, SFU and more ! 
      - Both MIT and Windows variations support
  - NTLM: 
      - refactor, clean SSP 
      - remove relay (TODO: reimplement with the new system)
  - Extensive GSSAPI / SPNEGO support !
  - Ticketer++ 
      - CCache support: read/write 
      - Kerberos integration: ask/renew/resign/edit tickets 
      - Change ticket fields through a GUI ! and more !
  - LDAP 
      - Fixes, ASN.1 Windows variation support 
      - dclocator, answering machine for "LDAP PING"
  - Automaton: 
      - fixes (memory usage on Windows) 
      - support for EOF events 
      - spawn() mode, better socket.socket support and more !
  - StreamSocket changes, support for TCP reassembly, etc !
  - Unit tests for everything (using samba, etc.)

**How to review**:
- build the doc locally and browse SMB/Kerberos/DCERPC
- try out `smbclient` and `smbserver` against Windows machines
- look at the code

fixes https://github.com/secdev/scapy/issues/4170

### CODE REVIEW

**Comment by gpotter2** [comment]
Ready for review !

**Comment by polybassa** [comment]
I reviewed a few files that I have an idea about, and everything looks good so far

**Review by guedou — APPROVED** [review]
Awesome PR!

File: doc/scapy/layers/dcerpc.rst [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    )
+
+
+.. note:: Currently, a DCERPC_Server will let a client bind on all interfaces that Scapy has registered (imported). Supposedly though, you know which RPCs are going to be queried.
+
+
+Define custom packets
+---------------------
+
+TODO: Add documentation on how to define NDR packets.
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply doc/scapy/layers/dcerpc.rst:341]: I mean there's still no doc regarding how to define NDR packets. There are enough examples though.

File: scapy/layers/http.py [hunk 2]
```diff
@@ -715,7 +714,6 @@ def http_request(host, path="/", port=80, timeout=3,
     try:
         ans = sock.sr1(
             req,
-            session=TCPSession(app=True),
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/layers/http.py:718]: Yeah. `StreamSession` now applies `tcp_reassemble` if available (as `TCPSession(app=True)` did).

File: doc/scapy/layers/dcerpc.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (guedou) [reply doc/scapy/layers/dcerpc.rst:341]: I think that you can remove this TODO.

[the remark this answers is not in the document]
Comment (gpotter2) [reply doc/scapy/layers/dcerpc.rst:341]: I'd rather keep it as a remainder. I'll probably end up adding some more doc eventually.

File: test/scapy/layers/kerberos.uts [hunk 3]
```diff
@@ -676,12 +789,175 @@ pkt = PACTYPE(
...
+
+print(outp)
+
+assert outp == """
+Tickets:
+0. Administrator@DOMAIN.LOCAL -> krbtgt/DOMAIN.LOCAL@DOMAIN.LOCAL
+   canonicalize+pre-authent+initial+renewable+proxiable+forwardable
+Start time         End time           Renew until        Auth time        
+27/09/22 17:29:30  28/09/22 03:29:30  30/09/22 17:29:27  27/09/22 17:29:30
+""".strip()
```
Comment (evverx) [inline test/scapy/layers/kerberos.uts:928]: This test seems to be failing with non-UTC timezones:
```sh
$ timedatectl show --property Timezone
Timezone=Europe/Berlin
```
```python
######                                                                                                                                                        
## Ticketer++ tests                                                                                                                                           
######                                                                                                                                                        
                                                                                                                                                              
                                                                                                                                                              
###(029)=[failed] Ticketer++ - Call show()                                                                                                                    
                                                                                                                                                              
>>> with ContextManagerCaptureOutput() as cmco:                                                                                                               
...     t.show()                                                                                                                                              
...     outp = cmco.get_output().strip()                                                                                                                      
...                                                                                                                                                           
>>> print(outp)                                                                                                                                               
Tickets:                                                                                                                                                      
0. Administrator@DOMAIN.LOCAL -> krbtgt/DOMAIN.LOCAL@DOMAIN.LOCAL                                                                                             
   canonicalize+pre-authent+initial+renewable+proxiable+forwardable                                                                                           
Start time         End time           Renew until        Auth time                                                                                            
27/09/22 19:29:30  28/09/22 05:29:30  30/09/22 19:29:27  27/09/22 19:29:30                                                                                    
>>>                                                                                                                                                           
>>> assert outp == """                                                                                                                                        
... Tickets:                                                                                                                                                  
... 0. Administrator@DOMAIN.LOCAL -> krbtgt/DOMAIN.LOCAL@DOMAIN.LOCAL                                                                                         
...    canonicalize+pre-authent+initial+renewable+proxiable+forwardable                                                                                       
... Start time         End time           Renew until        Auth time                                                                                        
... 27/09/22 17:29:30  28/09/22 03:29:30  30/09/22 17:29:27  27/09/22 17:29:30                                                                                
... """.strip()                                                                                                                                               
Traceback (most recent call last):                                                                                                                            
  File "<input>", line 2, in <module>                                                                                                                         
AssertionError
```

**Pull request merged**